ANN: Fix custom macro attr (name conflicts)

Stzx opened the pull request:
https://github.com/intellij-rust/intellij-rust/pull/5582#discussion_r443247691

Stzx commented on src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt:
```diff
@@ -3456,4 +3456,52 @@ class RsErrorAnnotatorTest : RsAnnotatorTestBase(RsErrorAnnotator::class) {
+    //- main.rs
+        use dep_proc_macro::repr;
+
+        #[repr/*caret*/(C)]
```
The local test will fail (the error is still reported), am I missing something?

Undin commented on src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:
```diff
@@ -1190,3 +1191,8 @@ private fun checkTypesAreSized(holder: RsAnnotationHolder, fn: RsFunction) {
 }
+
+// When `resolvePath()` parses the value, it means this is a custom proc macro (It currently looks like this
+private fun RsAttr.checkMetaItemName(target: String): Boolean = (metaItem.name == target && metaItem.resolvePath() == null)
```
Maybe `isBuiltInAttribute` or something similar is better?

Undin commented on src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:
```diff
@@ -1190,3 +1191,8 @@ private fun checkTypesAreSized(holder: RsAnnotationHolder, fn: RsFunction) {
+// When `resolvePath()` parses the value, it means this is a custom proc macro (It currently looks like this
+private fun RsAttr.checkMetaItemName(target: String): Boolean = (metaItem.name == target && metaItem.resolvePath() == null)
+
+private fun RsMetaItem.resolvePath(): RsElement? = path?.reference?.resolve()
```
Do we really need it? Maybe inline?

Undin commented on src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:
```diff
@@ -844,13 +844,14 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
-                RsDiagnostic.IncorrectlyPlacedInlineAttr(metaItem.path?.referenceNameElement ?: metaItem, attr)
-                    .addToHolder(holder)
-            }
+        if (metaItem.name != "inline") return
```
Why do you not use `checkMetaItemName` here?

Stzx replied:
For attributes like `inline`, using 
```rust
//- dep-proc-macro/lib.rs
    use proc_macro::TokenStream;

    #[proc_macro_attribute]
    pub fn inline(attr: TokenStream, item: TokenStream) -> TokenStream {
        item
    }
//- main.rs
    extern crate dep_proc_macro;

    use dep_proc_macro::inline;

    #[inline] // ambiguity
    type Foo = i32;
```
directly will actually lead to `E0659`.

For other explicit statements, 
```rust
//- main.rs
    extern crate dep_proc_macro;

    #[dep_proc_macro::inline] // metaItem.name == null
    type Foo = i32;
```
```rust
//- main.rs
    extern crate dep_proc_macro;

    use dep_proc_macro::inline as test;

    #[test] // metaItem.name == test
    type Foo = i32;
```

the plug-in currently does not resolve the name into `inline`, so only `metaItem.name != "inline"` is used here. I think it is enough.

Undin replied:
> directly will actually lead to E0659.

Yep, I'm aware about it. Moreover, https://github.com/intellij-rust/intellij-rust/pull/5582#discussion_r443247691 says about it directly (see code example). But it's exactly what we want - we shouldn't try to provide any analysis if there is `E0659` for the particular attribute. Yep, currently, we don't handle `E0659` but it will be handled eventually

> the plug-in currently does not resolve the name into inline

Yeah, our name resolution doesn't work properly in this case as well. But it doesn't make your changes worse, i.e. your current changes fix some rare false-positive errors and I think it's ok that we don't handle all similar false-positive errors because of the bug in name resolution. But we don't provide other false-positive errors so it's definitely an improvement

Stzx commented:
I re-examined it. In fact, as described by https://github.com/intellij-rust/intellij-rust/pull/5641#discussion_r449950120, other built-in attributes have this phenomenon, so it seems that this problem does not need to be fixed.

@Undin what do you think?

Stzx commented:
@Undin When I cleaned and moved `RsPsiPattern.STD_ATTRIBUTES`, I found that some of the built-in properties no longer exist in the new version. Do I need to clean it or keep it? (E.g `link_args`...)

Stzx commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsAttr.kt:
```diff
@@ -8,15 +8,86 @@ package org.rust.lang.core.psi.ext
 }
 
+/**
+ * Source of attributes: [https://doc.rust-lang.org/1.44.1/reference/attributes.html#built-in-attributes-index]
```
I updated the list from `1.30.0` to `1.44.1` and removed some names that no longer exist in the new version.

Stzx commented:
ping @Undin

Stzx commented on src/main/kotlin/org/rust/lang/core/RsPsiPattern.kt:
```diff
@@ -25,64 +25,64 @@ object RsPsiPattern {
-        "no_std",
-        "recursion_limit",
-        "windows_subsystem",
+    val STD_ATTRIBUTES: Set<String> = setOf(
```
For built-in attribute that have been removed in the new version (E.g `linked_from`).  keep it, but it is better to provide detection only for toolchains with this attribute?

Undin replied:
I don't think that there is a big difference because we don't have any checks related to such attributes, so we can use any version of this list currently

Undin commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsAttr.kt:
```diff
@@ -20,3 +22,11 @@ val RsAttr.owner: RsDocAndAttributeOwner?
+val RsAttr.isBuiltInAttribute: Boolean
+    get() {
+        val name = metaItem.name ?: return false
+        if (name !in STD_ATTRIBUTES && name != "start") return false
```
Can we put `start` in `STD_ATTRIBUTES` list?

Stzx replied:
`start` currently exists in the unstable list, I think maybe it can?

Undin commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsAttr.kt:
```diff
@@ -20,3 +22,11 @@ val RsAttr.owner: RsDocAndAttributeOwner?
+        val name = metaItem.name ?: return false
+        if (name !in STD_ATTRIBUTES && name != "start") return false
+
+        return !this.hasInScope(name, MACROS)
```
`this` is redundant here

Stzx replied:
Done.

Undin commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsAttr.kt:
```diff
@@ -20,3 +22,11 @@ val RsAttr.owner: RsDocAndAttributeOwner?
         else -> error("Unsupported attribute type: $this")
     }
+
+val RsAttr.isBuiltInAttribute: Boolean
```
I think we should rename it somehow because in case of collision it's still builtin attribute

Stzx replied:
do you mean `metaItem.name`?

Undin commented on src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:
```diff
@@ -1208,3 +1208,5 @@ private fun checkTypesAreSized(holder: RsAnnotationHolder, fn: RsFunction) {
     }
 }
+
+private fun RsAttr.checkNameAndIsBuiltIn(target: String): Boolean = (metaItem.name == target && isBuiltInAttribute)
```
Personaly, I'd merge it with `isBuiltInAttribute` and name it `isBuiltinWithName` (or something similar)

Stzx replied:
Okay, I will merge it and move it to `RsAttr.kt`

Undin commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsAttr.kt:
```diff
@@ -20,3 +22,14 @@ val RsAttr.owner: RsDocAndAttributeOwner?
         else -> error("Unsupported attribute type: $this")
     }
+
+fun RsAttr.isBuiltinWithName(target: String): Boolean {
+    return if (metaItem.name != target) false
+    else {
+        val name = metaItem.name ?: return false
+
+        if (name !in STD_ATTRIBUTES) false
+        else
+            !hasInScope(name, MACROS)
+    }
+}
```
nit: I think we can slightly rewrite it to improve readability
```kotlin
fun RsAttr.isBuiltinWithName(target: String): Boolean {
    val name = metaItem.name ?: return false
    if (name != target) return false
    if (name !in STD_ATTRIBUTES) return false
    return !hasInScope(name, MACROS)
}
```

Also, I think we should move it into `RsErrorAnnotator.kt` file and make it private - unlikely we need it somewhere else

Stzx replied:
Done.

Stzx commented:
bors r+

Stzx commented:
bors r-

Stzx commented:
bors r=Undin
